src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -655,6 +695,9 @@ private synchronized void markIndexFailed(Index index)
                 SystemKeyspace.setIndexRemoved(baseCfs.keyspace.getName(), indexName);
 
             needsFullRebuild.add(indexName);
+
+            if (!index.getSupportedLoadTypeOnFailure().supportsWrites() && writableIndexes.remove(indexName) != null)
+                logger.info("Index [" + indexName + "] became not-writable.");
```

Comment:
@adelapena please notice I changed sthg from your path here. Specifically the queryable removal upon failure as it would fail [this](https://github.com/apache/cassandra-dtest/pull/69/files#diff-fb360a29e46831ca27d4bb8e77059503L359) which looks legit.

Reply:
I'm not sure this is inline with the `LoadType.NOOP` behaviour of the indexes. Indeed, it leaves `LoadType#supportsReads()` without usages. I think that, to preserve the behaviour of the default indexes, we should make a distinction between failures in the initial build task and failures in rebuilds. As I mention on the dtest review, the current behaviour would be:
```java
/**
 * Returns the type of operations supported by the index in case its building has failed and it's needing recovery.
 * 
 * @param isInitialBuild {@code true} if the failure is for the initial build task on index creation, {@code false}
 * if the failure is for a rebuild or recovery.
 */
default LoadType getSupportedLoadTypeOnFailure(boolean isInitialBuild)
{
    return isInitialBuild ? LoadType.WRITE : LoadType.ALL;
}
```

Reply:
Hi @adelapena thanks again for looking into this. You're raising the point I wanted to discuss with you once this ticket was merged: Is it reasonable a 'not ready' index should be serving reads or writes?

I think we're finding ourselves battling that, but that's a bigger discussion. Maybe in some other ticket we could recover my approach which iirc would defer to the index implementation what to do :shrug:

Reply:
> Is it reasonable a 'not ready' index should be serving reads or writes?

For initial build it should't serve reads, but for rebuilds it's more complicated because the index could still be in good condition. IMO if we were creating the index API from scratch or adding a new implementation probably the best option would be to set not properly (re)built index as unable to serve reads and writes. But, given that we have had the cf-based implementation around for a while, I think we should by now preserve their original behaviour and provide the API mechanisms to change that in new implementations and perhaps in the standard indexes in the future. That and the recovery task seems a pair of nice improvements to ship with this ticket.

Accepting reads and writes when failed makes some sense in the particular case of cf-based regular indexes because they don't have an initialization task and they are based on idempotent operations on the underlaying column family. I'd say they focus in availability over consistency because even failed rebuilds always leave the index is same or better condition than before. In contrast, it's easy to imagine other implementations where a failed rebuild leaves the index completely corrupt and unusable. I think there could easily be use cases out there relying on this behaviour, and I'm afraid of arbitrarily changing that behaviour without coming back. If we still want to move to the new approach, either in this ticket or in a dedicated one, we should probably open a discussion in the mail list to see if there are deployments relaying on the classic behaviour. 

Another tempting option to ease the migration of cf-based indexes to the new consistency-over-availability approach would be making the enum returned by `getSupportedLoadTypeOnFailure` configurable through index options, for example:
```
CREATE INDEX my_idx ON t WITH OPTIONS = {'supported_load_on_failed_init': 'WRITE', 'supported_load_on_failed_rebuild': 'ALL'}
```
This is something that perhaps we could do in the ticket to change the default behaviour of regular cf-based indexes, while keeping this ticket focused on the API changes and the recovery task.

WDYT? Maybe I'm wrong and there isn't anyone out there relying on the old 2i behaviour and we should go straight to the the new failed-rebuild-means-broken behaviour.

Reply:
Hi @adelapena, regarding `return isInitialBuild ? LoadType.WRITE : LoadType.ALL;`

`isInitialBuild` is a new variable we'd have to create and track in SIM. And then `Index` would have to render which load is supported depending on 'when' the failure happened?. Sounds weird to me and like we are mixing concerns. It's easier for my brain to pick the idea of having the supported load state in the `Index`, SIM to track that instead of trying to infer it, set the defaults we'd like and be done. But that's for another discussion as you rightfully point.

I did try that change in SIM and this sends `SIMTest` into the most amazing failure scenarios with indexes blocking when they shouldn't, jstaks on all threads blocked and all kind of funnies. I did spend some time on it without success until I realized it should be:

`return isInitialBuild ? LoadType.WRITE : LoadType.READ;`

Correct?

Reply:
Hi @bereng. I understand that the original behaviour was:
```java
return isInitialBuild ? LoadType.WRITE : LoadType.ALL;
```
That's because the indexes were always writable independently of any kind of build failures, there wasn't even a notion of writability. And, once they were queryable once, they were queryable forever.

Just took a quick look at the `SIMTest` failures, I think they are happening because `TestingIndex.shouldFailCreate` is wrongly missed from `TestingIndex.clear()`. Once it's fixed to reset it, and we change a couple of assertions about writability in `initializingIndexNotQueryableButWritableAfterPartialRebuild`, all the tests seem to pass, at least locally.

I'll take a closer look tomorrow.

Reply:
Ah good catch! you're right that is being missed. Amazing this didn't came up until today.

In any case I'd like your feedback on the extra flush also. Feel free to shoot a half-backed push if you fancy and when I get the rest of the feedback I'll adjusts tests, dtests and that should be it hopefully.

Reply:
I think that the flush should happed right after we have ensured that the indexes are marked as writable. Otherwise, we might have new untracked writes populating the new memtable while the indexes are not writable. In the same vein, the sstable selection should happen after we have marked the indexes as writable. I have almost ready a commit with this and a few minor suggestions, I'll push it tomorrow. 

We're almost there!